src/nnet3/nnet-am-decodable-simple.cc

```diff
@@ -203,15 +196,83 @@ void NnetDecodableBase::GetCurrentIvector(int32 output_t_start,
                 << " (mismatched --ivector-period?)";
     }
     ivector_frame = online_ivector_feats_->NumRows() - 1;
+  } else if (ivector_frame < 0) {
+    int32 margin = - ivector_frame;
+    if (margin * online_ivector_period_ > 50) {
+      // Half a second seems like too long to be explainable as edge effects.
+      KALDI_ERR << "Could not get iVector for frame " << frame_to_search
+                << ", only available till frame "
+                << "0 * ivector-period=" << online_ivector_period_
+                << " (mismatched --ivector-period?)";
+    }
+    ivector_frame = 0;
   }
-  *ivector = online_ivector_feats_->Row(ivector_frame);
+  ivector->CopyFromVec(online_ivector_feats_->Row(ivector_frame));
 }
 
+void NnetDecodableBase::GetIvectorsForFrames(int32 output_t_start,
+                                             int32 num_output_frames,
+                                             Matrix<BaseFloat> *ivectors) {
+  const int32 left_context = nnet_left_context_ + opts_.extra_left_context,
+        right_context = nnet_right_context_ + opts_.extra_right_context;
+  // frame_to_search is the frame that we want to get the most recent iVector
+  // for.  We choose a point near the middle of the current window, the concept
+  // being that this is the fairest comparison to nnet2.   Obviously we could do
+  // better by always taking the last frame's iVector, but decoding with
+  // 'online' ivectors is only really a mechanism to simulate online operation.
+  if (ivector_interval_ == 0) { // single ivector case
```

Comment:
If zero is indeed a special case then it should be documented where you
document  GetTModulusForIvector().

On Thu, Mar 10, 2016 at 8:35 PM, Vijayaditya Peddinti <
notifications@github.com> wrote:

> In src/nnet3/nnet-am-decodable-simple.cc
> <https://github.com/kaldi-asr/kaldi/pull/468#discussion_r55778771>:
>
> >  }
> >
> > +void NnetDecodableBase::GetIvectorsForFrames(int32 output_t_start,
> > +                                             int32 num_output_frames,
> > +                                             Matrix<BaseFloat> *ivectors) {
> > +  const int32 left_context = nnet_left_context_ + opts_.extra_left_context,
> > +        right_context = nnet_right_context_ + opts_.extra_right_context;
> > +  // frame_to_search is the frame that we want to get the most recent iVector
> > +  // for.  We choose a point near the middle of the current window, the concept
> > +  // being that this is the fairest comparison to nnet2.   Obviously we could do
> > +  // better by always taking the last frame's iVector, but decoding with
> > +  // 'online' ivectors is only really a mechanism to simulate online operation.
> > +  if (ivector_interval_ == 0) { // single ivector case
>
> ivector_interval_ is being set by a call to GetTModulusForIvector in the
> constructor. Does getting a zero here mean that this is a single ivector
> case ?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kaldi-asr/kaldi/pull/468/files#r55778771>.
>